messaging/src/main/java/org/frankframework/messaging/amqp/AmqpListenerContainer.java

```diff
@@ -0,0 +1,262 @@
+			session = connection.openSession(sessionOptions);
+		} catch (ClientException e) {
+			throw new LifecycleException("Cannot open AMQP connection", e);
+		}
+	}
+
+	public boolean isOpen() {
+		return connection != null;
+	}
+
+	/**
+	 * Open a listener in this listener-container. The listener-container should have an open connection
+	 * before a listener can be opened, and the listener connection-name should be for the same connection
+	 * as this listener-container instance.
+	 *
+	 * @param amqpListener
+	 */
+	public void openListener(@Nonnull AmqpListener amqpListener) {
+		if (this.connectionName == null) {
+			throw new IllegalStateException("connection for the AmqpListenerContainer is not yet open");
+		}
+		if (!this.connectionName.equals(amqpListener.getConnectionName())) {
+			throw new IllegalStateException("AmqpListener[" + amqpListener.getName() +
+					"] connection [" + amqpListener.getConnectionName() +
+					"] is not the same as connection [" + this.connectionName +
+					"] for this AmqpListenerContainer");
+		}
+		ReceiverOptions options = new ReceiverOptions();
+		options.deliveryMode(amqpListener.getDeliveryMode());
+		options.sourceOptions().capabilities(amqpListener.getAddressType().getCapabilityName());
+
+		Receiver receiver;
+		try {
+			if (amqpListener.isDurable()) {
+				receiver = session.openDurableReceiver(amqpListener.getAddress(), amqpListener.getSubscriptionName(), options);
+			} else {
+				receiver = session.openReceiver(amqpListener.getAddress(), options);
+			}
+			receivers.put(receiver, amqpListener);
+			listeners.put(amqpListener, receiver);
+		} catch (ClientException e) {
+			throw new LifecycleException(e);
+		}
+
+		if (!isRunning.get()) {
+			startListenerThread();
+		}
+	}
+
+	/**
+	 * Return true when last listener is closed
+	 *
+	 * @param amqpListener
+	 * @return
+	 */
+	public boolean closeListener(AmqpListener amqpListener) {
+		Receiver receiver = listeners.remove(amqpListener);
+		if (receiver != null) {
+			receivers.remove(receiver);
+			receiver.close();
+		} else {
+			log.warn("No AMQP Receiver found for listener[{}]", amqpListener.getName());
+		}
+		isClosing.set(receivers.isEmpty());
+		return receivers.isEmpty();
+	}
+
+	private void startListenerThread() {
+		Thread listenerThread = new Thread(() -> {
+			isRunning.set(true);
+			try (Connection ignored = connection; Session theSession = session) {
+				while (!isClosing.get()) {
+					final Receiver receiver = theSession.nextReceiver(5, TimeUnit.SECONDS);
+					if (receiver == null) {
+						continue;
+					}
+					AmqpListener listener = receivers.get(receiver);
+					Delivery delivery = receiver.receive();
+					processDelivery(listener, delivery);
+				}
+			} catch (ClientException e) {
+				log.warn("Error receiving messages", e);
+			} finally {
+				isRunning.set(false);
+				connection = null;
+			}
+		});
+		listenerThread.start();
+	}
+
+	private void processDelivery(AmqpListener listener, Delivery delivery) {
```

Comment:
Eigenlijk doe je hier wat de pullingListenerContainer doet toch? Vanwaar de afweging de listener 'pushing' te laten implementeren?

Reply:
(Ik heb in ieder geval een test met 3 verschillende addresses, 2x queue en 1x topic, allemaal op dezelfde Session en voor zover ik kan testen gaat dit goed. Berichten worden correct ontvangen. Ik heb niet getest met verschillende delivery-modes maar omdat dit een optie is op de Receiver en niet de Session verwacht ik hierbij ook geen probleem).

Reply:
Gaat `final Receiver receiver = theSession.nextReceiver(5, TimeUnit.SECONDS);` iteratief alle receivers af?
Mocht je dan 20 listeners hebben met de zelfde connectie, doet hij dan 5x 20 seconden?

Reply:
Voor zover ik weet is de timeout voor de `nextReceiver()` method zelf, en niet per receiver.